airflow/providers/tableau/operators/tableau.py

```diff
@@ -0,0 +1,140 @@
+    :type match_with: Optional[str]
+    :param site_id: The id of the site where the workbook belongs to.
+    :type site_id: Optional[str]
+    :param blocking_refresh: By default the extract refresh will be blocking means it will wait until it has finished.
+    :type blocking_refresh: bool
+    :param tableau_conn_id: The :ref:`Tableau Connection id <howto/connection:tableau>`
+        containing the credentials to authenticate to the Tableau Server.
+    :type tableau_conn_id: str
+    """
+
+    def __init__(
+        self,
+        *,
+        resource: str,
+        method: str,
+        find: str,
+        match_with: str = 'id',
+        site_id: Optional[str] = None,
+        blocking_refresh: bool = True,
+        tableau_conn_id: str = 'tableau_default',
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.resource = resource
+        self.method = method
+        self.find = find
+        self.match_with = match_with
+        self.site_id = site_id
+        self.blocking_refresh = blocking_refresh
+        self.tableau_conn_id = tableau_conn_id
+
+    def execute(self, context: dict) -> str:
+        """
+        Executes the Tableau API resource and pushes the job id or downloaded file URI to xcom.
+        :param context: The task context during execution.
+        :type context: dict
+        :return: the id of the job that executes the extract refresh or downloaded file URI.
+        :rtype: str
+        """
+
+        available_resources = RESOURCES_METHODS.keys()
+        if self.resource not in available_resources:
+            error_message = f'Resource not found! Available Resources: {available_resources}'
+            raise AirflowException(error_message)
+
+        available_methods = RESOURCES_METHODS[self.resource]
+        if self.method in available_methods:
+            error_message = f'Method not found! Available methods for {self.resource}: {available_methods}'
+            raise AirflowException(error_message)
+
+        with TableauHook(self.site_id, self.tableau_conn_id) as tableau_hook:
+
+            resource = getattr(tableau_hook.server, self.resource)
+            method = getattr(resource, self.resource)
+
+            resource_id = self._get_resource_id(tableau_hook)
+
+            response = method(resource_id)
+
+        if self.method == 'refresh':
+
+            job_id = response.id
+
+            if self.blocking_refresh:
+
+                finish_code = TableauJobFinishCode.PENDING
+                negative_codes = (TableauJobFinishCode.ERROR, TableauJobFinishCode.CANCELED)
+                while not finish_code == TableauJobFinishCode.SUCCESS:
+                    return_code = int(tableau_hook.server.jobs.get_by_id(job_id).finish_code)
+                    finish_code = TableauJobFinishCode(return_code)
+                    if finish_code in negative_codes:
+                        raise TableauJobFailedException(f'The Tableau Refresh {self.resource} Job failed!')
```

Comment:
The hook has `wait_for_state` function which was recently added. Why not using it?

Reply:
@eladkal nice. I changed to use that  now.